Sentry 24.10.0 ImportError: cannot import name 'UserEmail' from 'sentry.models'

Hello!

I ran in to a problem with Senty Self-Hosted 24.10.0.

Want to note that everything worked fine until 24.4.0, after upgrade getting ImportError.

From the container `web-1` log files:

```log
2024-10-18T11:52:24.784493835Z   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
2024-10-18T11:52:24.784498079Z   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
2024-10-18T11:52:24.784503328Z   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
2024-10-18T11:52:24.784506170Z   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
2024-10-18T11:52:24.784508842Z   File "<frozen importlib._bootstrap_external>", line 995, in exec_module
2024-10-18T11:52:24.784513996Z   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
2024-10-18T11:52:24.784516739Z   File "/.venv/lib/python3.12/site-packages/sentry_auth_ldap/backend.py", line 3, in <module>
2024-10-18T11:52:24.784522249Z     from sentry.models import (
2024-10-18T11:52:24.784525014Z ImportError: cannot import name 'UserEmail' from 'sentry.models' (/usr/src/sentry/src/sentry/models/__init__.py)
2024-10-18T11:52:24.784530846Z 11:52:24 [ERROR] django.request: Internal Server Error: /auth/login/sentry/ (status_code=500 request=<WSGIRequest: POST '/auth/login/sentry/'>)
2024-10-18T11:52:24.784716822Z 11:52:24 [INFO] sentry.access.api: api.access (method='POST' view='sentry.web.frontend.auth_organization_login.AuthOrganizationLoginView' response=500 user_id='None' is_app='None' token_type='None' is_frontend_request='True' organization_id='None' auth_id='None' path='/auth/login/sentry/' caller_ip='10.6.0.122' user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36' rate_limited='False' rate_limit_category='None' request_duration_seconds=0.049422502517700195 rate_limit_type='DNE' concurrent_limit='None' concurrent_requests='None' reset_time='None' group='None' limit='None' remaining='None')
2024-10-18T11:52:30.449055902Z 11:52:30 [INFO] sentry.access.api: api.access (method='GET' view='sentry.web.frontend.auth_organization_login.AuthOrganizationLoginView' response=200 user_id='None' is_app='None' token_type='None' is_frontend_request='True' organization_id='None' auth_id='None' path='/auth/login/sentry/' caller_ip='10.6.0.122' user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36' rate_limited='False' rate_limit_category='None' request_duration_seconds=0.047009944915771484 rate_limit_type='DNE' concurrent_limit='None' concurrent_requests='None' reset_time='None' group='None' limit='None' remaining='None')
```

Container `web-1` installed python modules:

```shell
# pip list | grep sentry
sentry                             24.10.0     /usr/src/sentry/src
sentry-arroyo                      2.16.5
sentry-auth-ldap                   23.6.1
sentry-kafka-schemas               0.1.111
sentry_ophio                       1.0.0
sentry-protos                      0.1.23
sentry-redis-tools                 0.1.7
sentry-relay                       0.9.2
sentry-sdk                         2.15.0
sentry-usage-accountant            0.0.10
```

Here is my `sentry.conf.py` config file LDAP section:

```python
#######################
# LDAP Authentication #
#######################

import ldap
from django_auth_ldap.config import LDAPSearch, GroupOfUniqueNamesType

# Fields which managed users may not modify through the Sentry accounts view
SENTRY_MANAGED_USER_FIELDS = ('email', 'first_name', 'last_name', 'password', )
# AUTH_LDAP_DEFAULT_EMAIL_DOMAIN = 'example.com'

AUTH_LDAP_SERVER_URI = 'ldap://ldap.example.com:389'
AUTH_LDAP_BIND_DN = 'CN=sentry_ldap_provider,OU=Technological,OU=Users,OU=D18,DC=example,DC=com'
AUTH_LDAP_BIND_PASSWORD ='SuperSecurePasswordWasHere'
AUTH_LDAP_USER_QUERY_FIELD = 'username'

AUTH_LDAP_USER_SEARCH = LDAPSearch(
    'OU=Users,OU=D18,DC=example,DC=com',
    ldap.SCOPE_SUBTREE,
    '(&(objectCategory=Person)(memberOf=CN=SENTRY_TEST_USER,OU=Technological,OU=Groups,OU=GRP1,DC=example,DC=com)(|(mail=%(user)s)(sAMAccountName=%(user)s)))',
)

AUTH_LDAP_USER_ATTR_MAP = {
    'username': 'samaccountname',
    'name': 'cn',
    'email': 'mail'
}

AUTH_LDAP_MAIL_VERIFIED = True

AUTH_LDAP_GROUP_SEARCH = LDAPSearch(
    '',
    ldap.SCOPE_SUBTREE,
    '(objectClass=groupOfUniqueNames)'
)

AUTH_LDAP_GROUP_TYPE = GroupOfUniqueNamesType()
AUTH_LDAP_REQUIRE_GROUP = None
AUTH_LDAP_DENY_GROUP = None
AUTH_LDAP_FIND_GROUP_PERMS = False
AUTH_LDAP_CACHE_GROUPS = True
AUTH_LDAP_GROUP_CACHE_TIMEOUT = 3600

# AUTH_LDAP_DEFAULT_SENTRY_ORGANIZATION = u'TEST Sentry' // old
AUTH_LDAP_SENTRY_DEFAULT_ORGANIZATION = 'sentry'
AUTH_LDAP_SENTRY_ORGANIZATION_ROLE_TYPE = 'member'
AUTH_LDAP_SENTRY_GROUP_ROLE_MAPPING = {
    'owner': ['sysadmins'],
    'admin': ['devleads'],
    'member': ['developers', 'seniordevelopers']
}
AUTH_LDAP_SENTRY_ORGANIZATION_GLOBAL_ACCESS = False

AUTHENTICATION_BACKENDS = AUTHENTICATION_BACKENDS + (
    'sentry.utils.auth.EmailAuthBackend',
    'sentry_auth_ldap.backend.SentryLdapBackend',
)

# Optional logging for diagnostics.
LOGGING['disable_existing_loggers'] = False
import logging
logger = logging.getLogger('django_auth_ldap')
logger.setLevel(logging.DEBUG)
```

Is there something wrong with my configuration? Or in the latest Sentry releases has something changed?

--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
 *.pyc
 sentry_auth_ldap.egg-info
 dist
+.DS_Store

--- a/sentry_auth_ldap/backend.py
+++ b/sentry_auth_ldap/backend.py
@@ -6,21 +6,20 @@ from sentry import __version__ as sentry_version
 
 def compare_versions(current: str, required: str) -> bool:
     """
-    Compare two version strings.
+    Check if the current version meets the required version.
 
     :param current: The current version string
     :param required: The required version string
-    :return: True if the current version is greater than the required version, otherwise False
+    :return: True if the current version is greater than or equal to the required version, otherwise False
     """
-    return version.parse(current) > version.parse(required)
+    return version.parse(current) >= version.parse(required)
 
 # Import different models for backwards compatibility
-if compare_versions(sentry_version, "24.7.1"):
+if compare_versions(sentry_version, "24.8.0"):
     from sentry.models import Organization, OrganizationMember, UserOption
     from sentry.users.models import UserEmail
 else:
     from sentry.models import Organization, OrganizationMember, UserEmail, UserOption
-
 
 def _get_effective_sentry_role(ldap_user):
     role_priority_order = [

--- a/sentry_auth_ldap/backend.py
+++ b/sentry_auth_ldap/backend.py
@@ -1,11 +1,26 @@
 from django_auth_ldap.backend import LDAPBackend
 from django.conf import settings
-from sentry.models import (
-    Organization,
-    OrganizationMember,
-    UserEmail,
-    UserOption,
-)
+
+from packaging import version
+from sentry import __version__ as sentry_version
+
+def compare_versions(current: str) -> bool:
+    return version.parse(current) > version.parse("24.7.1")
+
+if compare_versions(sentry_version):
+    from sentry.models import (
+        Organization,
+        OrganizationMember,
+        UserOption,
+    )
+    from sentry.users.models import UserEmail
+else:
+    from sentry.models import (
+        Organization,
+        OrganizationMember,
+        UserEmail,
+        UserOption,
+    )
 
 
 def _get_effective_sentry_role(ldap_user):

--- a/sentry_auth_ldap/backend.py
+++ b/sentry_auth_ldap/backend.py
@@ -4,23 +4,22 @@ from django.conf import settings
 from packaging import version
 from sentry import __version__ as sentry_version
 
-def compare_versions(current: str) -> bool:
-    return version.parse(current) > version.parse("24.7.1")
+def compare_versions(current: str, required: str) -> bool:
+    """
+    Compare two version strings.
 
-if compare_versions(sentry_version):
-    from sentry.models import (
-        Organization,
-        OrganizationMember,
-        UserOption,
-    )
+    :param current: The current version string
+    :param required: The required version string
+    :return: True if the current version is greater than the required version, otherwise False
+    """
+    return version.parse(current) > version.parse(required)
+
+# Import different models for backwards compatibility
+if compare_versions(sentry_version, "24.7.1"):
+    from sentry.models import Organization, OrganizationMember, UserOption
     from sentry.users.models import UserEmail
 else:
-    from sentry.models import (
-        Organization,
-        OrganizationMember,
-        UserEmail,
-        UserOption,
-    )
+    from sentry.models import Organization, OrganizationMember, UserEmail, UserOption
 
 
 def _get_effective_sentry_role(ldap_user):

